VPC Service Controls demo pull request (#4)
terraform-google-modules/terraform-google-vpc-service-controls: examples/onprem_demo/README.md

We recommend that you delete the service account .json file and delete the service account altogether when you've completed running through this demo.

## Requirements
To complete this demo, you'll need:
- A Google Cloud [organization](https://cloud.google.com/resource-manager/docs/quickstart-organizations) mapped to a domain that you own with at least one project already created.
- A valid billing account where you have Billing Account User privileges.
- A service account with Project Creator privileges.  Steps for provisioning this are included below, but those steps assume you have Organization Admin privileges.
- Terraform v0.11.13 with provider.google v2.6.0 (the only versions I've tested; run `terraform version` to see what you've got).

## Instructions

0. Set up the service account needed to run the Terraform code.
    - In an existing project, create the service account.  Copy and paste its email address (the big, long, `gserviceaccount.com` address) to a scratch pad.  Download its .json secret key file into the `onprem_project/` and also copy it into the `vpc_sc_project` folders.  Be **extremely careful** with this file; delete it when you're done with this example.

>     - Go to the [IAM admin](https://console.cloud.google.com/iam-admin/iam) section of your *organization.*  (You should see your domain name at the top of the screen, not an individual project name).  Click Add.  In the "New Members" box, paste the email address of the service account you created.  Under the Roles dialog, search for "Project Creator."  Grant that role by clicking Save.  Your service account can now create projects and act as a project owner for those projects.

Reviewer: ```suggestion
    - Go to the [IAM admin](https://console.cloud.google.com/iam-admin/iam) section of your *organization.*  (You should see your domain name at the top of the screen, not an individual project name).  Click Add.  In the "New Members" box, paste the email address of the service account you created.  Under the Roles dialog, search for "Project Creator."  Grant that role by clicking Save.  Your service account can now create projects and act as a project owner for those projects.
    -  Add the additional Organization level permissions to the service account mentioned [here](https://github.com/terraform-google-modules/terraform-google-vpc-service-controls#organization-level-permissions) to  make sure your service account can create VPC SC otherwise known as Access Context Manager resources.
```

Author: Done.

>     - Go to the Billing section of the Cloud Console.  Select the billing account you want to charge for the projects you create.  (If you're using your credit card, the default limit is 5 projects mapped to a billing account, in case project creation fails for you in the steps below).  On the right side of the screen, there's a "Permissions" section for your billing account.  Click "Add members," and paste the email address of the service account you created.  Grant the service account Billing Admin privileges, so that it can map the projects it creates to your billing account.  **Note the billing account ID** (the alphanumeric text separated by hyphens), as we'll use it below.

Reviewer: ```suggestion
    - Go to the Billing section of the Cloud Console.  Select the billing account you want to charge for the projects you create.  (If you're using your credit card, the default limit is 5 projects mapped to a billing account, in case project creation fails for you in the steps below).  On the right side of the screen, there's a "Permissions" section for your billing account.  Click "Add members," and paste the email address of the service account you created.  Grant the service account Billing Account User privileges, so that it can map the projects it creates to your billing account.  **Note the billing account ID** (the alphanumeric text separated by hyphens), as we'll use it below.
```

Reviewer: Change from Billing Admin to Billing Account User

Author: Done.

1. Create the Terraform variables file.

>     - In your favorite text editor, edit and save the following variables in `examples/onprem_demo/terraform.tfvars`:

Reviewer: ```suggestion
    - In your favorite text editor, create and save a file named `terraform.tfvars` in the `examples/onprem_demo/` directory:
```

Author: Done.

        - `vpc_sc_project_id`: the unique ID for the VPC Service Control project that is going to be created.  Use lowercase letters and a 6 digit random number only (no spaces).
        - `onprem_project_id`: the unique ID for the on-prem project that is going to be created.  Use lowercase letters and a 6 digit random number only (no spaces).
        - `credentials_path`: the name of the service account secret key file you downloaded.  Relative path is fine if you saved it in the `onprem_project` directory.
        - `organization_id`: run `gcloud organizations list` to get this **number**.
        - `billing_account_id`: the billing account that the project will charge to.
        - `region`: region where the GCP resources will be created. 
        - `vpn_shared_secret`: any random text string can go here.

>         - `access_policy_name`: Get this value by running the following two commands:

Reviewer: As per this [comment](https://github.com/terraform-google-modules/terraform-google-vpc-service-controls/pull/4#discussion_r301367739) you should add a step to create the policy as a user may not have used VPC SC before. Or maybe a conditional step like "if `gcloud access-context-manager policies list --organization <Organization ID> ` returns nothing, create the policy"

Author: Done.

            - `gcloud organizations list` to get the Organization ID
            - `gcloud access-context-manager policies list --organization <Organization ID>` 
2. From a command line terminal:
    - `cd examples/onprem_demo`

>     - `terraform plan`

Reviewer: ```suggestion
    - `terraform init`
    - `terraform plan`
```

Author: Done.

    - `terraform apply` and type "yes" when it prompts you to create the resources.  Might take a few minutes to create the projects (should be less than 5 minutes).
    - If you get error messages about the Compute Engine API or Cloud DNS API needing to be enabled, wait a minute and try running `terraform apply` again.  If you get error messages about forwarding rules and VPN tunnel creation, wait a minute and try again.

To gain access to a VM in the VPC Service Control protected environment:

- Set the Windows password for the jumphost in the on-prem project, and download the .rdp file
- Set the Windows password for the VPC SC Windows VM
- RDP into the jumphost using its public IP
- Once you're in the jumphost, click Start -> Remote Desktop, and then RDP into `10.7.0.2` (the internal IP of the VM in the VPC Service Controls perimeter).  So you RDP into the jumphost, and then you RDP through the VPN tunnel into your cloud VM.  
- Play around with VPC Service Controls.
- Remember to delete all resources when done.  Easiest way to do this is to run `terraform destroy`.

> When you're done with this demo, delete the service account keys and delete the service account altogether from the Cloud Console.

Reviewer: Adding these placeholders are necessary to create documentation for your example.
Once you've added them, install terraform-docs and run `make generate_docs` command from the root folder as described [here](https://github.com/terraform-google-modules/terraform-google-vpc-service-controls#autogeneration-of-documentation-from-tf-files)

```suggestion
When you're done with this demo, delete the service account keys and delete the service account altogether from the Cloud Console.

[^]: (autogen_docs_start)
[^]: (autogen_docs_end)
```

Author: Is there any way we can avoid this?  This seems like needless complexity.  Much easier for me to simply type some text in a README than to figure out what's going on in that autogen link.

Reviewer: No unfortunately this is using terraform-docs is our standard way of documenting all modules and examples in modules.
If you're having trouble making it work let me know. It shouldn't be harder than downloading terraform-docs, adding it to your path and running `make generate_docs`

Author: Done.
